### ISSUE TITLE
PEP 712: `int` converter doesn't allow non-int arguments

### ISSUE THREAD

**Issue by thejcannon** [microsoft/pyright#5071]
**Describe the bug**
Using `int` as a `converter` in `dataclasses.field` and providing different non-int values results in errors

**To Reproduce**
```python
import dataclasses

from typing import cast

@dataclasses.dataclass
class Foo:
    converted: int = dataclasses.field(converter=int)

# These error
Foo("1")
Foo(cast(str, "1"))
Foo(1.0)

class TurnsIntoInt:
    def __int__(self):
        return 0
    

Foo(TurnsIntoInt())
```

**Expected behavior**
These should all be supported cases, and are likely to be very real-world examples (`str` and `int` come to mind)

**VS Code extension or command-line**
```
$ pyright --version
pyright 1.1.306
```

Using `pyright path/to/file.py`

**Additional context**
Thanks for the support. I'm still amazed and flattered.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

(@debonte, this is my bad. When I sent you an outline of how to implement the PEP 712 support, I forgot about the need to handle constructors. It's trivial to add.)

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.307, which I just published. It will also be included in a future release of pylance.
